### PR TITLE
ADDING ChargeItem resource with required HL7 fields & Verily enriched fields (hardcoded)

### PR DESCRIPTION
**Pull request by ivanBum**
Adding required fields
1. Satus: 1..1	code - ChargeItemStatus (ex. billable)
2.  Code: 1..1	CodeableConcept - A code that identifies the charge, like a billing code ChargeItemCode
3. Subject: 1..1	Reference(Patient | Group)	Individual service was done for/to

### CODE REVIEW

**Comment by ivanBum** [comment]
Adding verily enriched resources
1. enrichedRevenueCode
2. revenueCategory
3. enrichedHCPCS
4. detailedCost

It is important to mention that the values are hardcoded, meaning that they are dummy data for the purpose of testing the output.

**Comment by ivanBum** [comment]
**Adding output references and Fhir Store**

1. Output References
full patient bundle:
https://drive.google.com/file/d/1n-30IHd6C-Ct-nQ--8b9JMg93Z0Mqowj/view?usp=sharing
only chargeItem resource part of the bundle (for simplicity):
https://drive.google.com/file/d/1C7MMFgXafNrr84aNwv33A1t_-JGnv85d/view?usp=sharing

2. Fhir Store:
projects/value-suite-bumaschny-dev/locations/us-central1/datasets/synthea/fhirStores/verilyIG
The fhir store has the VerilyIG implemented, but I'm not sure it is actually working 
image

**Comment by ivanBum** [comment]
Questions I would need to know in order to continue:
1. What profile do I use? I currently have the HL7 us-core, and if I change it to verilyIG the fhir store rejects it (that's why I think the validation is not working propperly)
2. Is it correct to use the enriched fields verily dataStructure? Or do I model the chargeItem resource fully as the HL7 structure?
3. Do we need to change the values (form dummy to "real)? If so, then do we code them at the exporter level (as a function that adds random data to the chargeItem fields) or do we change it at the person generator (adding the information at the lowest level possible, probably best practice but really laborous)

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
+import org.hl7.fhir.r4.model.ChargeItem;
+
 // Duration to set length of stay
 import java.util.concurrent.TimeUnit;
 import org.hl7.fhir.r4.model.Duration;
 
-import ca.uhn.fhir.context.FhirContext;
 import ca.uhn.fhir.model.dstu2.valueset.EncounterStateEnum;
 
+import com.oracle.truffle.js.runtime.objects.Null;
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java]: Not sure if this package is used! If so, consider removing it.

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 2]
```diff
@@ -1,18 +1,28 @@
 package org.mitre.synthea.export;
 
+// Add chargeItem resource
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java]: The comment is not needed, consider removing it. The package is imported so the CahrgeITem is added and comment does not carry extra meaning.

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 3]
```diff
@@ -347,6 +361,11 @@ public static Bundle convertToFHIR(Person person, long stopTime) {
 
       explanationOfBenefit(personEntry, bundle, encounterEntry, person,
           encounterClaim, encounter, encounter.claim);
+      
+      if (USE_VERILY_EXTENSIONS) {
+      // one chargeItem per encounter
+      chargeItem(person, personEntry, bundle, encounterEntry, encounter);
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java]: Add indent please.

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 4]
```diff
@@ -178,19 +190,21 @@ public class FhirR4 {
...
   protected static final String SYNTHEA_IDENTIFIER = "https://github.com/synthetichealth/synthea";
+
+  // Verily Extension URL
   public static final String BASE_VVS_EXTENSION_URL = "https://verily-src.github.io/vhp-hds-vvs-fhir-ig/StructureDefinition/";
-  
+
   final static String LENGTH_OF_STAY_SUFFIX = "length-of-stay";
 
+  // Add Verily extension flag
+  protected static boolean USE_VERILY_EXTENSIONS =
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java]: Consider making it final.

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 5]
```diff
@@ -1465,6 +1484,98 @@ private static BundleEntryComponent explanationOfBenefit(BundleEntryComponent pe
...
+
+    // enrichedHCPCS
+    Extension enrichedHCPCS = codeEnrichment(ENRICHED_HCPCS, HCPCS_CODES);
+    chargeItemResource.addExtension(enrichedHCPCS);
+
+    // detailedCost
+    Extension moneyUrl = new Extension(DETAILED_COST);
+    Extension moneyExtension = new Extension("totalCost");
+    Money moneyResource = new Money();
+    moneyResource.setValue(4928.71);
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java:1555]: Is this always supposed to constant?

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 6]
```diff
@@ -1465,6 +1484,98 @@ private static BundleEntryComponent explanationOfBenefit(BundleEntryComponent pe
     return newEntry(person, bundle,eob);
   }
 
+  /**
+   * Map the ChargeItem into a FHIR ChargeItem resource, and add it to the given Bundle.
+   *
+   * @param rand           Source of randomness to use when generating ids etc
+   * @param personEntry    The Entry for the Person
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java:1489]: Would you elaborate what's person? Is it Patient? Or a Bundle Entry?

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 7]
```diff
@@ -3265,12 +3376,32 @@ public static void lengthOfStay(org.hl7.fhir.r4.model.Encounter encounterResourc
...
+
       // Once the Duration object is constructed and defined, create and extension and add this information
       String url = losExtensionUrl;
       Extension lengthOfStay = new Extension(url);
       lengthOfStay.setValue(length_of_stay);
       encounterResource.addExtension(lengthOfStay);
     } 
   }
+
+  public static Extension codeEnrichment(String fieldUrl, String codeUrl) {
```
Comment (vmontagh) [inline src/main/java/org/mitre/synthea/export/FhirR4.java:3395]: Consider adding JavaDoc.

**Review by vmontagh — COMMENTED** [review]
Overall LG, but added a few comments to improve the code. Also:

1. Would you change the base of the PR to master. The PR is a diff from [chargeItem](https://github.com/ivanBum/synthea/tree/chargeItem). Are the changes of that branch reviewed?
1. Do you have a plan to add unit tests? It seems that [FHIRR4ExporterTest.java](https://github.com/ivanBum/synthea/blob/master/src/test/java/org/mitre/synthea/export/FHIRR4ExporterTest.java) is the place to add tests. It would be great to have both tests and changes in the same PR.
1. IIUC, the mandatory fields are populated and a few optional fields, such as  `moneyResource` is populated with constant value. Is the plan to populate them with constant values, some random values, or somehow get them from user-input?

File: src/main/java/org/mitre/synthea/export/FhirR4.java (hunk 5 above)
Comment (ivanBum) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:1555]: No. This falls under the same question you asked above about the randomize fields. I made the same comment a few days ago; At this stage, the values are all constant in order to test functionality, but we should review if it is important to make them randomized and _how_ could we do It. I mentioned this before, so I'll proceed to add a GDoc explaining possible solutions to be explored

File: src/main/java/org/mitre/synthea/export/FhirR4.java (hunk 6 above)
Comment (ivanBum) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:1489]: The best way to describe this is to go way back and explain how the whole software works.
1. A `Person` object is a full representation of a generated individual (or patient, in this case interchangeable) within the sythea state-machine workflow (meaning that this person has no standard or json or anything). `./run_synthea` calls a Generator object and gives it all input parameters  
2. `./run_synthea` -> `createPerson(seed, demographics)` -> `export(person, exportConfig)` 
3.  The exporter checks validity of the process, and if all good, calls `FhirR4.convertToFHIR(person)`. At this stage, the person will be transformed from the state-machine output to a .JSON format with FHIR standard
4. The `personEntry` is the mapping of the person to the `Patient` resource
5. At the `convertToFhir()` stage, a bundle is generated and aggregated with the resource; this means that within this function, resources get generated from the original person for every possible resource, from `condition` to `chargeItem`. Every resource entry that are to be included in the bundle are called `bundleEntry`
This means that a `chargeItem bundleEntry` is a creation of the chargeItem resource with Fhir standard that is proceeded to be added to the person's Bundle

Comment (ivanBum) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:1489]: I'll create a GDoc adding all this information and possible discussions regarding the chargeItem, the synthea output and verilyIG

File: src/main/java/org/mitre/synthea/export/FhirR4.java (hunk 7 above)
Comment (ivanBum) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:3395]: Adding this JavaDoc  
```
/**
   * This function creates an enrichment template with "mainCode" and "enrichedCode" and
   * fills it with the specific URL's regarding the extension in question 
   * 
   * @param fieldUrl  CodeSystem terminology of the extension
   * @param codeUrl   full URL of the field in question, describing its structure     
   * @return enrichment extension
   */
```

**Comment by ivanBum** [comment]
> Overall LG, but added a few comments to improve the code. Also:
> 
> 1. Would you change the base of the PR to master. The PR is a diff from [chargeItem](https://github.com/ivanBum/synthea/tree/chargeItem). Are the changes of that branch reviewed?
> 2. Do you have a plan to add unit tests? It seems that [FHIRR4ExporterTest.java](https://github.com/ivanBum/synthea/blob/master/src/test/java/org/mitre/synthea/export/FHIRR4ExporterTest.java) is the place to add tests. It would be great to have both tests and changes in the same PR.
> 3. IIUC, the mandatory fields are populated and a few optional fields, such as  `moneyResource` is populated with constant value. Is the plan to populate them with constant values, some random values, or somehow get them from user-input?

1. chargeItem branch is a clone of extension made in order to finish the chargeItem resource export without need of merging it to the extensions branch directly. It is curated and aproved, and the difference with master is the implementation of LengthOfStay and the verily flag (PR#2)
2. Creating a task to do so
3. Creating a task to describe this and other possible discussions

**Comment by ivanBum** [comment]
Doc with questions and answers given in this PR 
https://docs.google.com/document/d/1GWNY8aJ6gqAE8pUujvgZd2bo4RBGklLbi-_JCr1hRso/edit?resourcekey=0-CJYW6-RRqYWhtU5mM1AoRQ#heading=h.kw36wcklsnu8

**Pull request merged**